File: impl/python/writ/issue.py
"""Builders that construct and sign protocol objects.

Every builder assembles the object itself from parts it was given and signs
that. narrow() is the only way to make a child writ: it starts from a parent
writ the caller holds, applies overrides, checks attenuation, and signs. No
function here accepts a literal writ object to sign (section 7.5, 12).
"""

import os
import time

from . import bounds as B
from . import chain as C
from . import objects as O
from .canon import canonicalize
from .errors import WritError
from .keys import b64u_encode

hash_body = O.hash_body


def nonce():
    """16 random bytes, base64url, 22 characters."""
    return b64u_encode(os.urandom(16))


def _sign(body, key):
    """Sign a freshly assembled object. Private on purpose."""
    body = {k: v for k, v in body.items() if k != "sig"}
    return {**body, "sig": key.sign(O.signing_input(body))}


def issue_root(key, holder, bnd, exp, nnc=None):
    """Issue a root writ from key to holder with bounds bnd until exp."""
    writ = _sign({
        "v": 1, "typ": "writ", "iss": key.did, "hld": holder, "bnd": bnd,
        "prv": None, "exp": exp, "nnc": nnc or nonce(),
    }, key)
    return O.verify_object(writ, "writ")


def narrow(parent, key, holder, exp=None, bnd=None, nnc=None):
    """Issue a child of ``parent`` held by ``key``, to ``holder``.

    ``bnd`` maps bound names to replacement bounds; every parent bound not
    replaced is carried over unchanged, and names absent from the parent are
    added. ``exp`` defaults to the parent's. The child is verified as a
    valid child under section 4 before it is returned; a widening override
    raises not_narrowed and nothing is signed.
    """
    parent = O.verify_object(parent, "writ")
    if key.did != parent["hld"]:
        raise WritError("chain_broken", "only the parent's hld may issue a child")
    child_bnd = dict(parent["bnd"])
    for name, bound in (bnd or {}).items():
        B.validate(bound, name)
        if name in child_bnd and not B.narrows(bound, child_bnd[name]):
            raise WritError("not_narrowed", f"override for {name} does not narrow the parent")
        child_bnd[name] = bound
    child_exp = parent["exp"] if exp is None else exp
    if child_exp > parent["exp"]:
        raise WritError("not_narrowed", "child exp is after parent exp")
    body = {
        "v": 1, "typ": "writ", "iss": key.did, "hld": holder, "bnd": child_bnd,
        "prv": O.identity(parent), "exp": child_exp, "nnc": nnc or nonce(),
    }
    O.check_bnd(child_bnd)
    C.check_pair(parent, body)
    return O.verify_object(_sign(body, key), "writ")


def make_call(key, chain, op, args, call_id=None):
    """Sign a call from key under chain (root first)."""
    chain = [O.verify_object(w, "writ") for w in chain]
    call = _sign({
        "v": 1, "typ": "call", "id": call_id or nonce(), "chain": chain,
        "from": key.did, "op": op, "args": args,
    }, key)
    return O.verify_object(call, "call")


def make_tally(key, call, writ, st="ok", out=None, used=None, rev=None,
               sub=None, wrt=None, acc=None, err=None):
    """Sign the executor's tally for ``call`` under leaf ``writ``.

    ``out`` may be a hash string or a result body (any JSON value), in which
    case its canonical hash is committed. ``err`` defaults to null for ok
    and to {"code": "unknown_outcome"} otherwise.
    """
    call = O.check_structure(call, "call")
    writ = O.check_structure(writ, "writ")
    if out is not None and not isinstance(out, str):
        out = O.hash_body(out)
    if err is None and st != "ok":
        err = {"code": "unknown_outcome"}
    tally = _sign({
        "v": 1, "typ": "tally", "call": O.identity(call), "writ": O.identity(writ),
        "op": call["op"], "acc": int(time.time()) if acc is None else acc,
        "st": st, "err": err, "out": out, "used": used or {}, "rev": rev,
        "sub": sub or [], "wrt": wrt or [],
    }, key)
    return O.verify_object(tally, "tally", signer=key.did)


def make_revoke(key, target, chain=None):
    """Sign a revoke of writ ``target`` (a writ object) or of every writ ("*")."""
    if target == "*":
        body = {"v": 1, "typ": "revoke", "writ": "*", "iss": key.did, "chain": []}
    else:
        chain = [O.verify_object(w, "writ") for w in (chain or [])]
        body = {"v": 1, "typ": "revoke", "writ": O.identity(target), "iss": key.did, "chain": chain}
    return O.verify_object(_sign(body, key), "revoke")


def to_bytes(obj):
    """Canonical bytes of an object, for sending or storing."""
    return canonicalize(obj)

File: impl/go/writ/issue.go
package writ

import (
	"crypto/rand"
	"encoding/json"

	"writproto/keys"
	"writproto/wire"
)

func random22() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	return wire.B64.EncodeToString(b)
}

// Nonce is the nonce source; tests replace it for reproducible vectors.
var Nonce = random22

// toNumber converts Go integers in a freshly built object into json.Number so
// that the object matches what Decode would produce.
func normalize(obj wire.Object) (wire.Object, error) {
	b, err := json.Marshal(obj)
	if err != nil {
		return nil, err
	}
	return wire.Decode(b)
}

// Issue builds and signs a writ from iss to hld. When parent is non-nil the
// writ is a child of parent and MUST narrow it; Issue refuses otherwise, so an
// implementation cannot accidentally sign a widened grant. bnd is the JSON
// shape: name to {"t": type, "v": value}.
func Issue(iss *keys.Identity, hld string, bnd map[string]any, exp int64, parent *Writ) (*Writ, error) {
	obj := wire.Object{
		"v": 1, "typ": "writ", "iss": iss.DID(), "hld": hld,
		"bnd": bnd, "exp": exp, "nnc": Nonce(),
	}
	if parent == nil {
		obj["prv"] = nil
	} else {
		obj["prv"] = parent.ID
	}
	obj, err := normalize(obj)
	if err != nil {
		return nil, err
	}
	if err := wire.Sign(obj, iss); err != nil {
		return nil, err
	}
	w, err := ParseWrit(obj)
	if err != nil {
		return nil, err
	}
	if parent != nil {
		if err := CheckChild(w, parent); err != nil {
			return nil, err
		}
	}
	return w, nil
}

// NewCall builds and signs a call under chain.
func NewCall(from *keys.Identity, chain []*Writ, op string, args map[string]any) (*Call, error) {
	var raw []any
	for _, w := range chain {
		raw = append(raw, w.Raw)
	}
	if raw == nil {
		raw = []any{}
	}
	if args == nil {
		args = map[string]any{}
	}
	obj := wire.Object{
		"v": 1, "typ": "call", "id": Nonce(), "chain": raw,
		"from": from.DID(), "op": op, "args": args,
	}
	obj, err := normalize(obj)
	if err != nil {
		return nil, err
	}
	if err := wire.Sign(obj, from); err != nil {
		return nil, err
	}
	return ParseCall(obj)
}

// TallyInput is what an executor knows when it signs a tally.
type TallyInput struct {
	Call     *Call
	Acc      int64
	St       string
	ErrCode  string
	ErrRef   string
	Res      any // result body, nil when none
	Used     map[string]int64
	RevUntil *int64
	Sub      []*Tally
	Wrt      []*Writ
}

// NewTally builds and signs a tally as the leaf holder.
func NewTally(exe *keys.Identity, in TallyInput) (*Tally, any, error) {
	var errv any
	if in.ErrCode != "" {
		e := map[string]any{"code": in.ErrCode}
		if in.ErrRef != "" {
			e["ref"] = in.ErrRef
		}
		errv = e
	}
	var out any
	if in.Res != nil {
		h, err := HashResult(in.Res)
		if err != nil {
			return nil, nil, err
		}
		out = h
	}
	used := map[string]any{}
	for k, v := range in.Used {
		used[k] = v
	}
	var rev any
	if in.RevUntil != nil {
		rev = map[string]any{"until": *in.RevUntil}
	}
	sub := []any{}
	for _, s := range in.Sub {
		sub = append(sub, s.Raw)
	}
	wrt := []any{}
	for _, w := range in.Wrt {
		wrt = append(wrt, w.Raw)
	}
	obj := wire.Object{
		"v": 1, "typ": "tally", "call": in.Call.ID, "writ": in.Call.Leaf().ID,
		"op": in.Call.Op, "acc": in.Acc, "st": in.St, "err": errv, "out": out,
		"used": used, "rev": rev, "sub": sub, "wrt": wrt,
	}
	obj, err := normalize(obj)
	if err != nil {
		return nil, nil, err
	}
	if err := wire.Sign(obj, exe); err != nil {
		return nil, nil, err
	}
	t, err := ParseTally(obj, exe.DID())
	if err != nil {
		return nil, nil, err
	}
	return t, in.Res, nil
}

// NewRevoke builds and signs a revoke of the leaf of chain (or "*" when chain is empty).
func NewRevoke(iss *keys.Identity, chain []*Writ) (*Revoke, error) {
	raw := []any{}
	target := "*"
	for _, w := range chain {
		raw = append(raw, w.Raw)
	}
	if len(chain) > 0 {
		target = chain[len(chain)-1].ID
	}
	obj := wire.Object{"v": 1, "typ": "revoke", "writ": target, "iss": iss.DID(), "chain": raw}
	obj, err := normalize(obj)
	if err != nil {
		return nil, err
	}
	if err := wire.Sign(obj, iss); err != nil {
		return nil, err
	}
	return ParseRevoke(obj)
}

// CheckRevoke validates a parsed revoke per section 9.1.
func CheckRevoke(r *Revoke) error {
	if r.Writ == "*" {
		if len(r.Chain) != 0 {
			return fail(Malformed, "key-wide revoke must have an empty chain")
		}
		return nil
	}
	if err := VerifyChain(r.Chain); err != nil {
		return err
	}
	if r.Chain[len(r.Chain)-1].ID != r.Writ {
		return fail(TallyMismatch, "revoke names %s but chain leaf is %s", r.Writ, r.Chain[len(r.Chain)-1].ID)
	}
	if !Issuers(r.Chain)[r.Iss] {
		return fail(NoStanding, "revoker %s is not an issuer on the chain", r.Iss)
	}
	return nil
}
